python/pyarrow/tests/test_dataset.py

```diff
@@ -2582,6 +2600,38 @@ def test_open_dataset_from_fsspec(tempdir):
     assert dataset.schema.equals(table.schema)
 
 
+@pytest.mark.parquet
+@pytest.mark.s3
+def test_file_format_inspect_fsspec(s3_filesystem):
+    # https://issues.apache.org/jira/browse/ARROW-16413
```

Comment:
Can we simply use a fsspec local filesystem to avoid the S3 scaffolding?

Reply:
I was just trying exactly the same :)

Reply:
So it does work (I checked the test hangs with a local fs as well, before applying the fix), with the caveat that I need to manually construct the PyFileSystem with FSSpecHandler, because if you pass an actual fsspec local filesystem, we internally convert it into a arrow native local filesystem:

https://github.com/apache/arrow/blob/7a0f00c16e084d194ae53d209b33b809cfc8f2d5/python/pyarrow/fs.py#L109-L113

Reply:
That's actually also the reason that our nightly integration tests with dask didn't catch this, because we only were running the main parquet tests that didn't use S3 but only a local filesystem

Reply:
Hmm, actually, we can perhaps even remove `fsspec` out of the equation and instead use `PyFileSystem(ProxyHandler(LocalFileSystem()))`?